### DISCUSSION TITLE
Possibility of dynamic Repeater title

### DISCUSSION THREAD

**Discussion by kalkoen** [area17/twill#1079] [Q&A] [answered]
Is it possible to make a Repeater Title dynamic? E.g. make it a link to a more advanced edit page of that  item.
It seems to me that sometimes, you wish to have a Repeater like structure, i.e. a Story that has many Characters, but you might have so much information about one Character, that you prefer not to edit it all in the Story form. Instead, you want to add Characters on the Story page, but have the Repeater title be the link to the editor of that particular Character. So that you have one big edit page for that character.

I have considered using a Browser-based option, but in my case, the Story to Character is one-to-many, not many-to-many.

Is this possible, or should I be looking at a different way of achieving my goal, that does not involve a Repeater?

**Comment by pboivin** [comment, marked as answer]
Hey @kalkoen,

It's not currently possible to have a dynamic repeater title, but there's a PR on this that's been kicking around: https://github.com/area17/twill/pull/149

It's something I regularly find myself wishing for, so I'm thinking of working on this in the near future :)

To your question of linking to full-featured forms from repeater items — I believe this could be achieved with a simple link. If both modules are added to the admin routes and only 'story' is added to the CMS navigation, you could include a link in the 'character.blade.php' repeater template to take the user to the full 'character' edit form. The link would be something like: `/admin/characters/{$id}/edit`

Hope this can help!

**Reply by kalkoen** [reply]
Thanks for the reply! Dynamic repeater titles would be great indeed! 

As for simple links. Yes, that should indeed work. The only thing I'm struggling with is passing variabes to the role.blade.php. For a full form, I know I can inject variables in the module's Controller via formData, but how does this work when using a Repeater? It seems like CharacterController is not called when the Repeater is served in the Story form, so using CharacterController's formData is of no use.

**Reply by pboivin** [reply]
Not always a popular choice but personally, I often reach for [View::share()](https://laravel.com/docs/8.x/views#sharing-data-with-all-views) to pass data to nested Blade structures. Sharing from the right controller/action will limit the scope to specific pages.

Another option would be to fetch the data directly from the repeater template, something like:
```blade
@php
    $data = app(StoryRepository::class)->getMyData();
@endphp
```

**Reply by kalkoen** [reply]
Sounds good, thank you! 
It's getting quite offtopic now, so perhaps I'll make a post later explaining this situation, documenting it for anyone who might need it but:
I don't fully understand how Repeaters work in the background, because what I cannot get my head around is: in, character.blade.php, how can I use variables of that particular item in the repeater? i.e. create a link to that particular Character, in the Repeater item on the full Story field. It seems like there is not way of accessing that particular Character's variables in character.blade.php.

Perhaps a stupid attempt, but investigating `{{ dd(get_defined_vars()) }}` doesn't offer much help. In a full form field, formData in Controller would help. [This ](https://spectrum.chat/twill/help/access-variables-in-repeater~bcf07ba6-740d-43e0-9694-08f7aa831fda)discussion is a dead end too.

Finally, I am not sure how the two methods you mentioned above can help with adding such a link in the repeater, because I do not know how to 'identify' what Character I am dealing with in character.blade.php.

Thank you for your time and effort!

**Reply by pboivin** [reply]
Hey @kalkoen,

Sorry, I feel like I may have mixed a few different concerns in my thoughts on repeaters.

`View::share()` will probably be more useful when you are working with repeater items on your frontend. In the context of the CMS, it won't help you with your specific problem.

Reaching for `dd(get_defined_vars())` is far from stupid! Also, I understand that the output is probably confusing in the case of the repeater template...

A bit of context — Your `character.blade.php` is not the full story, like a regular Blade partial would be. In the context of repeaters and block editors, this Blade code actually generates a _block template_, that is pushed to the frontend, to then hydrate the blocks within the repeater field.

I may not be explaining this perfectly, but the point is:
1. your `character.blade.php` doesn't have access to any meaningful data out of the gate, and
2. it is used as a template for the Vue.js layer

To your actual problem — you could generate your link like this:
```vue
<a :href="'/admin/characters/'+$parent.block.id.replace(/\D*/, '')+'/edit'">Edit full character</a>
```

This is hacky and quite far from the intended useage... If anything, this points to an area that could be improved to enable more use cases with repeaters and blocks :)
